### DISCUSSION TITLE
$"{variable}" VS "{#}"

### DISCUSSION THREAD

**Discussion by kianzarrin** [CitiesSkylinesMods/TMPE#1484] [General]
![image](https://user-images.githubusercontent.com/26344691/159375207-0af28165-3747-417e-849b-ac6a79ad8dd1.png)
Is there any reason why we use Log._DebugFormat?
Following all those numbers can be confusing. its much easier if I just do 
```cs
                                    Log.Debug(
                                        $"segmentId:{segmentId} " + ...
```
this way I am less likely to make mistake.

**Comment by krzychu124** [comment]
The thing is that debug logs were created to analyze later in the log, not to read normally. I don't even like it was formatted as now these logs take 80% space of the code. It does not make it more readable, really - it was not meant to be readable.

**Comment by originalfoo** [comment]
There's several classes that are so extensively littered with those `_DebugFormat()` chunks that they obfuscate the code. Like, every few lines there's another massive chunk of debug logging.

Is there some way, even if painfully slow, that we could drastically reduce the bloat of debug logging in general?

**Reply by krzychu124** [reply]
> Is there some way, even if painfully slow, that we could drastically reduce the bloat of debug logging in general?

Create dedicated methods for each log call and pass variables? 😂 

No there is no other way.

**Reply by originalfoo** [reply]
Could something like this be used, but also query the value? https://stackoverflow.com/a/9801735

![image](https://user-images.githubusercontent.com/1386719/159379486-bd289fd3-ab0b-4200-8e94-0e8a0d752587.png)

It would create ton of gc, but could significantly reduce the bloat of debug logging, for example:

```csharp
Log._Dump(
   "arbitrary message string",
   () => someVar1,
   () => someVar2,
   // ...
);
```

**Reply by kianzarrin** [reply]
using $"" does reduce ossification too. but my main problem with 
```cs
                                    Log._DebugFormat(
                                        "RoutingManager.RecalculateLaneEndRoutingData({0}, {1}, {2}, {3}): " +
                                        "checking lane connections of nextLaneId={4}, idx={5}: " +
                                        "isNextStartNodeOfNextSegment={6}, nextSegmentId={7}, " +
                                        "nextHasConnections={8}, nextIsConnectedWithPrev={9}",
                                        segmentId,
                                        laneIndex,
                                        laneId,
                                        startNode,
                                        nextLaneId,
                                        nextLaneIndex,
                                        isNodeStartNodeOfNextSegment,
                                        nextSegmentId,
                                        nextHasConnections,
                                        nextIsConnectedWithPrev);
```
is that in order to determine `nextIsConnectedWithPrev={9}` prints the right variable, I have to manually count 9 parameters.

**Reply by kianzarrin** [reply]
Log._Dump sounds too complicated. I prefer something like this :
```cs
                                    Log._Debug(
                                        "prefer stay in lane information:\n" +
                                        $"prevSegmentId={prevSegmentId} prevLane:[id={prevLaneId} index={prevLaneIndex} outerSimilarLaneIndex:{prevOuterSimilarLaneIndex} similarLaneCount={prevSimilarLaneCount}]\n" +
                                        $"nextSegmentId={nextSegmentId} nextLane:[id={nextLaneId} index={nextLaneIndex} outerSimilarLaneIndex:{nextOuterSimilarLaneIndex} similarLaneCount={nextSimilarLaneCount}]\n" +
                                        $"nextIsTrackOnly={nextIsTrackOnly} similarLaneCountMatches={similarLaneCountMatches}  outerSimilarLaneIndexMatches={outerSimilarLaneIndexMatches} stayInlaneTracks={stayInlaneTracks}");
```

**Reply by kianzarrin** [reply]
@krzychu124 
> Is there some way, even if painfully slow, that we could drastically reduce the bloat of debug logging in general?
> 
> Create dedicated methods for each log call and pass variables? 😂
> 
> No there is no other way.

I created macro in VS to collapse all the logs:
```
to collapse all if (log...) {
select text and press Control+F
go to beggining of the doc
load macro: F3 -> Ctrl+M,M  
replay macro: Control+Shift+P many times
```
Not sure if Kider supports this
